docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+  possible, there will always be a delta to be addressed in Cluster API due to the fact that Cluster API is operated 
+  from the management cluster, while drain is done in workload clusters.
+
+  Additionally, Cluster API should take care of many clusters, while Kubernetes features are by design scoped to a single cluster.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### `MachineDrainRule` CRD
+
+We propose to introduce a new `MachineDrainRule` namespace-scoped CRD. The goal of this CRD is to allow management 
+cluster admins/users to define a drain rule (a specific drain behavior) for a set of Pods.
+
+Additionally, drain rules can be restricted to a specific set of Machines, e.g. targeting only controlplane Machines, 
+Machines of specific Clusters, Linux Machines, etc.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: example-rule
+  namespace: default
+spec:
+  drain:
+    # behavior can be Drain or Skip
+    behavior: Drain 
+    # order can only be set if behavior == Drain
+    # Pods with higher order are drained later
+    order: 100
+  machines:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    clusterSelector:
+      matchLabels:
+      matchExpressions:
+  pods:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    namespaceSelector:
+      matchLabels:
+      matchExpressions:
+```
+
+Notes:
+
+* All selectors are of type `metav1.LabelSelector`
+* In order to allow expressing selectors with AND and OR predicates, both `machines` and `pods` have a list of 
+  selectors instead of a single selector
+
+##### Example: Exclude Pods from drain
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: skip-pods
+  namespace: default
+spec:
+  drain:
+    behavior: Skip
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchExpressions:
+      - key: app
+        operator: In
+        values:
+        - example-app1
+        - example-app2
+        - example-app3
+    namespaceSelector:
+      matchLabels:
+        kubernetes.io/metadata.name: example-namespace
+```
+
+##### Example: Drain order
+
+(related to [CAPI-11024](https://github.com/kubernetes-sigs/cluster-api/issues/11024))
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: portworx
+  namespace: default
+spec:
+  drain:
+    behavior: Drain
+    order: 100
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchLabels:
+        app: portworx
+```
+
+#### Drain annotations
+
+We propose to introduce new annotations to allow workload cluster admins/users to define drain behavior 
+for Pods. These annotations would be either added directly to Pods or indirectly via Deployments, StatefulSets, etc.
+The annotations will take precedence over `MachineDrainRules` specified in the management cluster.
+
+* `cluster.x-k8s.io/drain: skip`
+* `cluster.x-k8s.io/drain-order: <order>`
```

Comment:
These are OK to support, but maybe we can define an annotation that `kubectl` (in client side drain mode) is also likely to care about.

Also, I'd use a label for these so that clients get to use selectors if they want to.

Reply:
> These are OK to support, but maybe we can define an annotation that kubectl (in client side drain mode) is also likely to care about. eg maintenance.node.kubernetes.io/skip-on-drain: "true"?

Not sure if you're suggesting to define this annotation instead of the CAPI-specific one or in addition 😀

In general in CAPI I think we can only define annotations in our Namespace.

If we define this one in k/k additionally I'm absolutely in favor of supporting it in CAPI

> Also, I'd use a label for these so that clients get to use selectors if they want to.

Good point, let's use labels instead

Reply:
> Drain _order_ is problematic. What happens if two Pods have the same ordinal? Maybe we mean drain priority?

We had priority initially, but it felt unintuitive to us. Basically does high drain priority mean it should be drained earlier or later? Order seemed clearer in that regard.

If multiple Pods have the same order they are drained "at the same time", i.e. within the same Reconcile and there is no guarantee that one gets evicted before the other

Reply:
WRT to label/annotation, I think we should start with CAPI specific ones, gather some feedback for one or two cycle. Only then eventually we should propose labels/annotation in k/k (and possibly do the work in kubectl drain).

WRT to k/k, another reason to wait a little bit before proposing new labels/annotations is to wait for more progress in the NodeMaintenance discussion (even if my current feeling is that this will take some time more, see e.g. https://github.com/kubernetes/enhancements/pull/4213#discussion_r1784080253)

Reply:
> Also, I'd use a label for these so that clients get to use selectors if they want to.

Done